File: 2asir/web-deployment/PHP/clase/OBJETOS/Python_POO_Mastery_Kit/01_Fundamentos.py
"""
01_Fundamentos.py

OBJETIVO:
Entender los bloques básicos de la Programación Orientada a Objetos (POO) en Python
(equivalente al PHP anterior).

CONCEPTOS:
1. Clase (class): La plantilla. En Python se definen con `class Nombre:`.
2. Objeto: La instancia.
3. Atributos (Propiedades): Variables ligadas a `self`.
4. Métodos: Funciones dentro de la clase que SIEMPRE reciben `self` como primer parámetro.
5. __init__: El constructor (equivalente a __construct de PHP).
"""

## -----------------------------------------
## SECCIÓN 1: DEFINICIÓN DE LA CLASE
## -----------------------------------------

class Libro:
    # En Python no se suelen declarar las propiedades fuera del constructor
    # se inicializan dinámicamente, aunque es buena práctica anotarlas.

    def __init__(self):
        # Inicializamos atributos vacíos
        self.titulo = ""
        self.autor = ""
        self.paginas = 0

    # MÉTODOS
    # OJO: En Python, 'this' se llama 'self' y hay que ponerlo EXPLICITAMENTE
    # como primer argumento.
    def mostrar_info(self):
        return f"El libro '{self.titulo}' escrito por {self.autor} tiene {self.paginas} páginas."

## -----------------------------------------
## SECCIÓN 2: INSTANCIACIÓN
## -----------------------------------------

print("--- CREANDO OBJETOS ---")

# En Python NO se usa 'new'. Se llama a la clase como si fuera una función.
libro1 = Libro()
# Accedemos con punto (.) en vez de flecha (->)
libro1.titulo = "El Principito"
libro1.autor = "Antoine de Saint-Exupéry"
libro1.paginas = 96

libro2 = Libro()
libro2.titulo = "1984"
libro2.autor = "George Orwell"
libro2.paginas = 328

## -----------------------------------------
## SECCIÓN 3: USANDO LOS OBJETOS
## -----------------------------------------

print(f"Libro 1: {libro1.titulo}")
print(f"Libro 2: {libro2.titulo}")

print("\n--- INFORMACIÓN DETALLADA ---")
print(libro1.mostrar_info())
print(libro2.mostrar_info())

"""
REFLEXIÓN DOCENTE COMPARATIVA:
- PHP: $this->variable
- Python: self.variable
- PHP: function nombre()
- Python: def nombre(self):
"""

File: 2asir/web-deployment/PHP/clase/OBJETOS/PHP_POO_Mastery_Kit/01_Fundamentos.php
<?php
/**
 * 01_Fundamentos.php
 * 
 * OBJETIVO:
 * Entender los bloques básicos de la Programación Orientada a Objetos (POO) en PHP.
 * La POO nos permite organizar el código modelando cosas del mundo real.
 * 
 * CONCEPTOS:
 * 1. Clase (Class): El plano o plantilla (ej. "Coche").
 * 2. Objeto (Object): Una instancia concreta (ej. "Mi Ferrari Rojo").
 * 3. Propiedades: Variables dentro de una clase (características).
 * 4. Métodos: Funciones dentro de una clase (comportamientos).
 * 5. Instanciación: Crear un objeto con `new`.
 */

## -----------------------------------------
## SECCIÓN 1: DEFINICIÓN DE LA CLASE
## -----------------------------------------

// Definimos la plantilla "Libro"
class Libro
{
    // PROPIEDADES (Características del libro)
    // 'public' significa que se pueden acceder desde fuera de la clase
    public $titulo;
    public $autor;
    public $paginas;

    // MÉTODOS (Acciones que puede hacer el libro o que se hacen con él)
    public function mostrarInfo()
    {
        // $this hace referencia a "este" objeto específico que estamos usando
        return "El libro '{$this->titulo}' escrito por {$this->autor} tiene {$this->paginas} páginas.";
    }
}

## -----------------------------------------
## SECCIÓN 2: INSTANCIACIÓN (CREAR OBJETOS)
## -----------------------------------------

echo "--- CREANDO OBJETOS ---\n";

// Creamos el primer objeto (una copia de la plantilla Libro)
$libro1 = new Libro();
$libro1->titulo = "El Principito";
$libro1->autor = "Antoine de Saint-Exupéry";
$libro1->paginas = 96;

// Creamos un segundo objeto totalmente independiente
$libro2 = new Libro();
$libro2->titulo = "1984";
$libro2->autor = "George Orwell";
$libro2->paginas = 328;

## -----------------------------------------
## SECCIÓN 3: USANDO LOS OBJETOS
## -----------------------------------------

// Accedemos a sus propiedades
echo "Libro 1: " . $libro1->titulo . "\n";
echo "Libro 2: " . $libro2->titulo . "\n";

// Llamamos a sus métodos
echo "\n--- INFORMACIÓN DETALLADA ---\n";
echo $libro1->mostrarInfo() . "\n";
echo $libro2->mostrarInfo() . "\n";

/* 
 * REFLEXIÓN DOCENTE:
 * Fíjate cómo usamos la MISMA función `mostrarInfo()` para los dos objetos,
 * pero cada uno imprime sus PROPIOS datos. Esa es la magia de $this.
 * $this es un camaleón que se adapta al objeto que lo llama.
 */
?>